Add SYCL Sort KT documentation to library guide (#2650)
uxlfoundation/oneDPL: documentation/library_guide/kernel_templates/sycl/radix_sort.rst

<https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_sycl_device_memory_model>`_)
for intermediate data storage. For the algorithm to operate correctly, there must be enough memory on the device.
If there is not enough global device memory, a ``std::bad_alloc`` exception is thrown.
The behavior is undefined if there is not enough local memory.
The amount of memory that is required depends on input data and configuration parameters, as described below.

Global Memory Requirements
--------------------------

Global memory is used for copying the input sequence(s) and storing internal data such as radix value counters.
The amount used depends on many parameters; below is an upper bound approximation:

   N\ :sub:`keys` + C * N\ :sub:`keys`

where the sequence with keys takes N\ :sub:`keys` space, and the additional space is C * N\ :sub:`keys`.

The value of `C` depends on ``param.data_per_workitem``, ``param.workgroup_size``, and ``RadixBits``.
For ``param.data_per_workitem`` set to `10`, ``param.workgroup_size`` to `512`, and ``RadixBits`` to `8`,
`C` is typically less than `1`.
Incrementing ``RadixBits`` increases `C` up to twice, while doubling either

> ``param.data_per_workitem`` or ``param.workgroup_size`` leads to a halving of `C`.

Reviewer: Does it make sense to talk about what changes to `RadixBits` does if only `8` is allowed currently?

Author: I agree it may be confusing since it makes it seem it is configurable. I also updated esimd docs to remove this mention.

   The ranking takes approximately "(workgroup_size / sub_group_size) * (2^RadixBits) * sizeof(uint16_t)" bytes.
   It surpasses Intel Data Center GPU / BMG Max SLM capacity in only marginal cases,
   e.g., when RadixBits is 10 and workgroup_size is 64, or when RadixBits is 9 and workgroup_size is 128.
   It is ignored as an unrealistic case.

-----------------------------------------
Recommended Settings for Best Performance
-----------------------------------------

The general advice is to choose kernel parameters based on performance measurements and profiling information.
The initial configuration may be selected according to these high-level guidelines:

- When the number of elements to sort ``N`` is small (e.g., less than ~1M), utilizing all available
  compute cores is key for better performance. Allow creating enough work chunks to feed all
  X\ :sup:`e`-cores [#fnote2]_ on a GPU: ``param.data_per_workitem * param.workgroup_size ≈ N / xe_core_count``.

- When the number of elements to sort is large (e.g., more than ~1M), maximizing the number of elements
  processed by a work-group, which equals to ``param.data_per_workitem * param.workgroup_size``,

>   reduces synchronization overheads between work-groups and usually benefits the overall performance.

Reviewer: Does it make sense to give examples of parameters for best performance for BMG and PVC for a couple sizes?  Maybe we don't want to get that specific, but I could see it being a convenient starting point.

Author: I think providing the best scalable performance values is a good place to start. This also ends up being a good upper-bound for dpwi. In the long-run, I wonder if some sort of database of experimental results makes sense.

I've listed sample parameters based on good performance settings for 2^28 uint32_t key / val types. We could list more, but it gets complex and differs between architectures. We probably want to think of some more structured way to provide these across all kernel templates.

Reviewer: I don't think we need to get too complicated here, I think a starting point is probably enough, at least for now. I can just see a user reading the previous docs and not knowing where to start tuning.

.. warning::

>    Avoid setting too large ``param.data_per_workitem`` and ``param.workgroup_size`` values.

Reviewer: Are we able to say more about what might be "too large"?

Author: Yes, I listed sample parameters that scale well. Going beyond these will lead to worsened performance and / or crashes due to device resource constraints.

   Make sure that :ref:`Memory requirements <sycl-radix-sort-memory-requirements>` are satisfied.

.. warning::

   Maximizing ``param.data_per_workitem`` generally improves performance as long as private memory usage does not exceed available register capacity.

>    Large performance drops with an increase to ``param.data_per_workitem`` is indicative of register spillage and profiling tools may be used to analyze this behavior.

Reviewer: ```suggestion
   Large performance drops with an increase to ``param.data_per_workitem`` are indicative of register spillage and profiling tools may be used to analyze this behavior.
```

Radix Sort
##########

-----------------------------
radix_sort Function Templates
-----------------------------

The ``radix_sort`` function sorts data using the radix sort algorithm.
The sorting is stable, preserving the relative order of equal elements.
Both in-place and out-of-place overloads are provided. Out-of-place overloads do not alter the input sequence.

The functions implement a Onesweep* [#fnote1]_ algorithm variant.

.. note::
   ``radix_sort`` is currently available for Intel® Arc™ B-Series and Intel® Data Center GPU Max Series.
   The Intel® oneAPI DPC++/C++ Compiler 2025.1.0 compiler or greater is required, and the Unified Runtime adapter over

>    Level Zero must be used. This is the default adapter for Intel GPUs. For more information, please refer to

Reviewer 2: ```suggestion
   The Intel® oneAPI DPC++/C++ Compiler 2025.1.0 or greater is required, and the Unified Runtime adapter over
   Level Zero must be used (which is the default adapter for Intel GPUs). For more information, please refer to
```

Radix Sort
##########

-----------------------------
radix_sort Function Templates
-----------------------------

The ``radix_sort`` function sorts data using the radix sort algorithm.
The sorting is stable, preserving the relative order of equal elements.
Both in-place and out-of-place overloads are provided. Out-of-place overloads do not alter the input sequence.

The functions implement a Onesweep* [#fnote1]_ algorithm variant.

.. note::
   ``radix_sort`` is currently available for Intel® Arc™ B-Series and Intel® Data Center GPU Max Series.
   The Intel® oneAPI DPC++/C++ Compiler 2025.1.0 compiler or greater is required, and the Unified Runtime adapter over

>    Level-Zero must be used. This is the default adapter for Intel GPUs. For more information, please refer to

Reviewer 2: ```suggestion
   Level Zero must be used. This is the default adapter for Intel GPUs. For more information, please refer to
```

   `DPC++ Device Selection

>    <https://www.intel.com/content/www/us/en/docs/oneapi/programming-guide/2025-1/device-selection.html>`_.

Reviewer 2: I think we need to add the link to `_auxiliary\common_cross_document_links.txt`

Author: Done

    +---------------------------+----------------------------+----------------------------+
    | Intel Arc B-Series        | 10                         | 512                        |
    +---------------------------+----------------------------+----------------------------+

.. note::

   Optimal parameters may differ by the data type and the entropy of the underlying data.
   Furthermore, smaller input sizes may perform better with smaller ``param.data_per_workitem`` values or a different
   ``param.workgroup_size``.

.. note::

   Avoid setting too large ``param.data_per_workitem`` and ``param.workgroup_size`` values.
   Make sure that :ref:`Memory requirements <sycl-radix-sort-memory-requirements>` are satisfied.
   If sorting 4-byte types, a good starting point is to use the ``param.data_per_workitem`` values in the sample
   parameters above as an upper-bound for experimentation.

.. note::

   Maximizing ``param.data_per_workitem`` generally improves scalable performance as long as private memory usage does not
   exceed available register capacity. Large performance drops with an increase to ``param.data_per_workitem`` are

>    indicative of register spillage and profiling tools may be used to analyze this behavior.

Reviewer 2: I do not think any of those should be highlighted as a note. Maybe it makes sense to format these as a list of tips (with some of the current notes possibly split into 2-3 independent tips).

Author: I tried a few ways out and I think a bulleted list within a `.. tip` directive looks the best (instead of several `.. tip` directives). I took the suggestion and broke these into several tips.

I moved the note discussing the table before the directive.

.. [#fnote1] Andy Adinets and Duane Merrill (2022). Onesweep: A Faster Least Significant Digit Radix Sort for GPUs. https://arxiv.org/abs/2206.01784.
.. [#fnote2] The X\ :sup:`e`-core term is described in the `oneAPI GPU Optimization Guide

>    <https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/2024-0/intel-xe-gpu-architecture.html#XE-CORE>`_.

Reviewer 2: The link to the GPU optimization guide is versioned, so better also added to the special file, and updated to the most recent version of the page.

Author: Done, I saw a place in the scan KT does this so I moved it to as well.

   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename Range>
   sycl::event
   radix_sort (sycl::queue q, Range&& r, KernelParam param = {}); // (2)

   // Sort out-of-place
   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename Iterator1,
             typename Iterator2>
   sycl::event
   radix_sort (sycl::queue q, Iterator1 first, Iterator1 last,
               Iterator2 first_out, KernelParam param = {}); // (3)

   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename Range1, typename Range2>
   sycl::event
   radix_sort (sycl::queue q, Range1&& r, Range2&& r_out,
               KernelParam param = {}); // (4)
   }

.. note::
   ``radix_sort`` is currently available for Intel® Arc™ B-Series and Intel® Data Center GPU Max Series.
   The Intel® oneAPI DPC++/C++ Compiler 2025.1.0 compiler or greater is required, and the Unified Runtime adapter over
   Level-Zero must be used. This is the default adapter for Intel GPUs. For more information, please refer to
   `Device Selection

>    <https://www.intel.com/content/www/us/en/docs/oneapi/programming-guide/2025-1/device-selection.html>`_.

Reviewer 2: ```suggestion
   Level-Zero must be used. This is the default adapter for Intel GPUs. For more information, please refer to
   `DPC++ Device Selection
   <https://www.intel.com/content/www/us/en/docs/oneapi/programming-guide/2025-1/device-selection.html>`_.
```

> 

Reviewer 2: I suggest to move the whole note above the synopsis.

Author: Done

|                                               |   <viewable-ranges>` (2,4).                                         |
|                                               |                                                                     |
|                                               |                                                                     |
|                                               |                                                                     |
+-----------------------------------------------+---------------------------------------------------------------------+
| ``param``                                     | A :doc:`kernel_param <../kernel_configuration>` object.             |
|                                               |                                                                     |
|                                               |                                                                     |
+-----------------------------------------------+---------------------------------------------------------------------+

**Type Requirements**:

- The element type of sequence(s) to sort must be a C++ integral or floating-point type

>   other than ``bool`` with a width of up to 64 bits.

Reviewer 2: A side question: is `sycl::half` supported? If not, why not? IIRC, our "main" radix sort implementation supports that.

Author: I had only added overloads for the order preserving casts from the key-type -> unsigned integer that were supported by ESIMD, so `sycl::half` is not currently supported.

There is no algorithmic limitation as to why this should be the case, so extending support should be easy. I opened https://github.com/uxlfoundation/oneDPL/issues/2664.

-----------------------------------------
Recommended Settings for Best Performance
-----------------------------------------

The general advice is to choose kernel parameters based on performance measurements and profiling information.
The initial configuration may be selected according to these high-level guidelines:

- When the number of elements to sort ``N`` is small (e.g., less than ~1M), utilizing all available
  compute cores may improve performance. Experiment with creating enough work chunks to feed all
  X\ :sup:`e`-cores [#fnote2]_ on a GPU: ``param.data_per_workitem * param.workgroup_size ≈ N / xe_core_count``
  in addition to the next configuration which maximizes data per work-group. The optimal settings may differ between
  hardware depending on the cost of inter-work group synchronization.

- When the number of elements to sort is large (e.g., more than ~1M), maximizing the number of elements
  processed by a work-group, which equals to ``param.data_per_workitem * param.workgroup_size``,
  reduces synchronization overheads between work-groups and usually benefits the overall performance.

Sample Parameters
-----------------

> 

Reviewer 2: This subheading does not look very useful to me, if not even the opposite as it breaks the section into two parts. If you would like to have a title for the table, consider the `.. table::` directive (https://docutils.sourceforge.io/docs/ref/rst/directives.html#tables)

Author: Thanks for the suggestion. I switched to use the `.. table::` directive.

+---------------------------+----------------------------+----------------------------+
| Platform                  | ``data_per_workitem``      | ``workgroup_size``         |
+===========================+============================+============================+
| Intel Data Center GPU Max | 22                         | 512                        |
+---------------------------+----------------------------+----------------------------+
| Intel Arc B-Series        | 10                         | 512                        |
+---------------------------+----------------------------+----------------------------+

.. note::

   Optimal parameters may differ by the data type and the entropy of the underlying data.
   Furthermore, smaller input sizes may perform better with smaller ``param.data_per_workitem`` values or a different
   ``param.workgroup_size``.

.. warning::

   Avoid setting too large ``param.data_per_workitem`` and ``param.workgroup_size`` values.
   Make sure that :ref:`Memory requirements <sycl-radix-sort-memory-requirements>` are satisfied.
   A good starting reference is to avoid using ``param.data_per_workitem`` values beyond those listed in the sample

>    parameters table above.

Reviewer 2: This tip sounds strange to me, or maybe I misunderstand. The table mentions just one value per device type. If any other value is better to be avoided, why is that a parameter at all? That just does not sound right.

Author: I think my wording was confusing. From [this discussion](https://github.com/uxlfoundation/oneDPL/pull/2650#discussion_r3060218913) I decided to add some sample configuration for the most common sort type we've seen that can serve as an example of the upper-bound for the tuning space. These are not recommended values for all configurations, but we can expect performance to significantly worsen when tuning beyond these values for the listed hardware and eventually out of resource errors.

We need a better way to recommend tuning values in the future for different sizes and across different types for all kernel templates.

.. warning::

   Maximizing ``param.data_per_workitem`` generally improves scalable performance as long as private memory usage does not
   exceed available register capacity. Large performance drops with an increase to ``param.data_per_workitem`` are

>    indicative of register spillage and profiling tools may be used to analyze this behavior.

Reviewer 2: Please take another look and think whether all that should be formatted as notes or warnings. Most of that seems to be general performance-related advice, in line with the goal of the subsection.

Author: It is performance guidance, so a note is probably better here. I made the change across the different files.

Reviewer 2: In some cases, maybe even a note is not needed and a regular text would be sufficient. I will take another look.
